File: modules/react-native-nitro-recycler-list/android/src/main/java/com/margelo/nitro/recyclerlist/HybridRecyclerCellHostView.kt
package com.margelo.nitro.recyclerlist

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.margelo.nitro.NitroModules
import com.margelo.nitro.views.RecyclableView

class HybridRecyclerCellHostView(
  context: Context = requireNotNull(NitroModules.applicationContext),
) : HybridRecyclerCellHostViewSpec(), RecyclableView {
  override val view = RecyclerCellContainer(context)

  override var listId: String = ""
  override var slotId: Double = -1.0
  override var itemKey: String = ""
  override var itemType: String = "default"

  private var previousListId = ""
  private var previousSlotId = -1

  override fun afterUpdate() {
    val nextSlotId = slotId.toInt()
    if (previousListId != listId || previousSlotId != nextSlotId) {
      if (previousListId.isNotEmpty()) RecyclerListRegistry.unregisterHost(this)
      previousListId = listId
      previousSlotId = nextSlotId
      RecyclerListRegistry.registerHost(this)
    }
  }

  fun addReactChild(child: View, index: Int) {
    (child.parent as? ViewGroup)?.removeView(child)
    view.addView(
      child,
      index.coerceIn(0, view.childCount),
      FrameLayout.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.WRAP_CONTENT,
      ),
    )
  }

  fun removeReactChild(child: View) {
    view.removeView(child)
  }

  override fun prepareForRecycle() {
    RecyclerListRegistry.unregisterHost(this)
    previousListId = ""
    previousSlotId = -1
    view.clearAnimation()
    view.translationX = 0f
    view.translationY = 0f
    view.alpha = 1f
    view.clearFocus()
    listId = ""
    slotId = -1.0
    itemKey = ""
    itemType = "default"
  }

  override fun onDropView() {
    RecyclerListRegistry.unregisterHost(this)
    (view.parent as? ViewGroup)?.removeView(view)
  }
}

File: modules/react-native-nitro-recycler-list/ios/HybridRecyclerCellHostView.swift
import NitroModules
import UIKit

final class HybridRecyclerCellHostView: HybridRecyclerCellHostViewSpec, RecyclableView {
  let view = RecyclerCellContainer()
  var listId: String = ""
  var slotId: Double = -1
  var itemKey: String = ""
  var itemType: String = "default"

  private var previousListId = ""
  private var previousSlotId = -1

  override init() {
    super.init()
    view.onComponentViewMounted = { [weak self] in
      guard let self else { return }
      RecyclerListRegistry.reconcile(host: self)
    }
  }

  func afterUpdate() {
    let nextSlotId = Int(slotId)
    guard previousListId != listId || previousSlotId != nextSlotId else { return }
    if !previousListId.isEmpty {
      RecyclerListRegistry.unregister(
        host: self,
        listId: previousListId,
        slotId: previousSlotId
      )
    }
    previousListId = listId
    previousSlotId = nextSlotId
    RecyclerListRegistry.register(host: self)
  }

  func prepareForRecycle() {
    RecyclerListRegistry.unregister(host: self)
    previousListId = ""
    previousSlotId = -1
    view.layer.removeAllAnimations()
    view.transform = .identity
    view.alpha = 1
    view.resignFirstResponder()
    listId = ""
    slotId = -1
    itemKey = ""
    itemType = "default"
  }

  func onDropView() {
    RecyclerListRegistry.unregister(host: self)
    view.onComponentViewMounted = nil
    view.removeFromSuperview()
  }
}
